src/messaging/ReliableMessageManager.cpp

```diff
@@ -413,18 +459,31 @@ CHIP_ERROR ReliableMessageManager::SendFromRetransTable(RetransTableEntry * entr
 
     if (rc)
     {
-        // Locally store the start and length;
-        uint8_t * p  = entry->msgBuf->Start();
-        uint16_t len = entry->msgBuf->DataLength();
+        System::PacketBufferHandle msgBuf;
+
+        msgBuf.Adopt(entry->msgBuf);
+
+        // Retain entry->msgBuf in mRetransTable for following retransmition
+        msgBuf->AddRef();
 
-        // Send the message through
-        uint16_t msgSendFlags = entry->msgSendFlags;
-        SetFlag(msgSendFlags, MessageFlagValues::kMessageFlag_RetainBuffer);
-        err = SendMessage(rc, entry->msgBuf, msgSendFlags);
+        uint32_t msgId         = 0;
+        uint32_t payloadLength = 0;
 
-        // Reset the msgBuf start pointer and data length after sending
-        entry->msgBuf->SetStart(p);
-        entry->msgBuf->SetDataLength(len);
```

Comment:
This is no longer needed because `SendMessage` does it now?  That seems like a pretty important part of the `SecureSessionMgr::SendMessage` API, but as far as I can tell it is NOT documented to have this behavior.  If all consumers of `SecureSessionMgr::SendMessage` want this behavior, it's fine to move it there, but that needs to be documented.  If it's just this consumer, the behavior should stay here...

Reply:
This logic is not valid in chip since Transport layer will do the in place encryption, we need to rewind the p right before the encryption portion of the payload  after sending the packet.

Reply:
Yes, I understand that now.  We still need to document this rewinding behavior in the API documentation for `SecureSessionMgr::SendMessage`.

Reply:
Added more comments in the body, this operation is handled inside the SendMessage implementation.

Reply:
At risk of being repetitive: This needs to be documented as part of the API documentation for `SecureSessionMgr::SendMessage` (which needs to start existing).

Reply:
Done.

Reply:
now that I'm thinking about this a bit more, I think I'd rather that SecureSession do the retention (i.e. return an object that it makes, which can be a PacketBufferHandle Subclass).

Something like:
```
SecureSession::SendMessage(... SecurePacketBufferHandle *retained_for_resend = nullptr);
SecureSession::SendMessage(SecurePacketBufferHandle &already_encrypted);
```

Basically: enforce that SecureSessionManager's encryption can't be bypassed by using the type system.